File: qc-application/src/main/java/com/eager/questioncloud/application/business/payment/question/implement/QuestionPaymentCouponProcessor.java
package com.eager.questioncloud.application.business.payment.question.implement;

import com.eager.questioncloud.core.domain.coupon.infrastructure.repository.UserCouponRepository;
import com.eager.questioncloud.core.domain.payment.model.QuestionPayment;
import com.eager.questioncloud.core.exception.CoreException;
import com.eager.questioncloud.core.exception.Error;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class QuestionPaymentCouponProcessor {
    private final UserCouponRepository userCouponRepository;

    public void applyCoupon(QuestionPayment questionPayment) {
        if (!questionPayment.checkUsingCoupon()) {
            return;
        }

        questionPayment.useCoupon();

        if (!userCouponRepository.use(questionPayment.getQuestionPaymentCoupon().getUserCouponId())) {
            throw new CoreException(Error.FAIL_USE_COUPON);
        }
    }
}

File: qc-api/qc-payment-api/src/main/kotlin/com/eager/questioncloud/payment/question/implement/QuestionPaymentCouponProcessor.kt
package com.eager.questioncloud.payment.question.implement

import com.eager.questioncloud.common.exception.CoreException
import com.eager.questioncloud.common.exception.Error
import com.eager.questioncloud.coupon.infrastructure.repository.UserCouponRepository
import com.eager.questioncloud.payment.domain.QuestionPayment
import org.springframework.stereotype.Component

@Component
class QuestionPaymentCouponProcessor(
    private val userCouponRepository: UserCouponRepository
) {
    fun applyCoupon(questionPayment: QuestionPayment) {
        if (!questionPayment.isUsedCoupon()) {
            return
        }

        questionPayment.applyCoupon()

        if (!userCouponRepository.use(questionPayment.questionPaymentCoupon!!.userCouponId)) {
            throw CoreException(Error.FAIL_USE_COUPON)
        }
    }
}
